### PR TITLE
Grand refactoring of instances lifecycle

### LINKED ISSUE #218: Join instance without polling membership
Initially algorithm of bootstrapping new instance looked as follows:

1. Check current workdir, find no snapshots in there.
2. Start polling config from a random member in membership.
3. If current advertise_uri is mentioned in config - start bootstrap process.

Now there is another option - we could use `cartridge.remote-control module` to perform netbox calls even before box.cfg was initialized. It'll make bootstrapping algorithm simpler and faster, and also will allow to handle some errors.

This is what should be done:

1. cartgrige.cfg should call remote_control.start()
2. bootstrap_from_membership should be removed
2. config_patch_clusterwide should trigger netbox.call bootstrap_from_scratch

### PR DESCRIPTION
**Pull request by rosik**
Changelog

- Whole new way to bootstrap instances. Instead of polling membership
  for getting clusterwide config the instance now start Remote Control
  Server (with limited iproto protocol functionality) on the same port.
  Two-phase commit is then executed over net.box connection.

- Module `cartridge.confapplier` is divided into several smaller ones
  with better isolation and more clear responsibilities: `roles`,
  `failover`, `twophase`.

- Module `cartridge.twophase` absorbed the code related to the
  two-phase commit algorithm. The main change here is the way new
  instances are joined: now they are the part of the quorum and do
  perform prepare_2pc and commit_2pc stages.

- Module `cartridge.roles` consolidates role management functions:
  `register_role`, some getters, `validate_config` and `apply_config`.
  The module is almost stateless, it's only state is a collection of
  registered roles. Roles state is managed with the `service-registry`
  as before.

- Module `cartridge.confapplier` itself implements internal state
  machine which helps to manage cluster operation and protects from
  invalid state transitions.

- Module `cartridge.bootstrap` is removed. Its responsibilities are
  inherited by confapplier.

- Function `cartridge.bootstrap` is removed. Use `edit_topology` instead.

- New internal module `cartridge.failover` is responsible for making
  decisions on instance leaderspip. It provides `is_rw` and `is_leader`
  methods to get current instance properties. It also monitors
  membership updates and reconfigure roles when needed.

- Failover triggered on instance status `dead` instead of `suspect`.

- Module `cartridge.topology` refactored to be stateless.
  Now it only provides methods for config validation and some popular
  transformations (`find_server_by_uri`, `get_fullmesh_replication`).
  The rest of responsibilities are now handled by other modules: active
  topology is managed by the confapplier, membership state is managed
  by failover module.

I did'n forget about

- [x] Tests
- [x] Changelog
- [x] Documentation

Close #218, #41, #33, #66, #190, #247, #228

### CODE REVIEW

File: cartridge/twophase.lua [hunk 1]
```diff
@@ -148,128 +187,140 @@ local function _clusterwide(patch)
...
+                log.error('Error aborting config update at %s: %s', uri, err)
             end
         end
+
+        goto finish
     end
 
+::finish::
     if _2pc_error == nil then
         log.warn('Clusterwide config updated successfully')
```
Comment (sharonovd) [inline cartridge/twophase.lua]: Should it be logged after abort?

File: cartridge/twophase.lua [hunk 2]
```diff
@@ -148,128 +187,140 @@ local function _clusterwide(patch)
...
-                conn,
-                '_G.__cluster_confapplier_prepare_2pc',
-                {conf_new}, {timeout = 5}
-            )
-            if ok == true then
+    table.sort(uri_list)
+
+    goto prepare
+
+::prepare::
```
Comment (sharonovd) [inline cartridge/twophase.lua]: Was goto really necessary?

File: cartridge/webui/api-topology.lua [hunk 3]
```diff
@@ -1,8 +1,8 @@
 #!/usr/bin/env tarantool
 
 local admin = require('cartridge.admin')
-local roles = require('cartridge.roles')
 local gql_types = require('cartridge.graphql.types')
+local roles = require('cartridge.roles')
```
Comment (sharonovd) [inline cartridge/webui/api-topology.lua]: Why did you move this line?

File: cartridge/confapplier.lua [hunk 4]
```diff
@@ -16,362 +20,452 @@ local pool = require('cartridge.pool')
...
 -- @treturn[2] nil
 -- @treturn[2] table Error description
-local function load_from_file(filename)
-    checks('?string')
-    filename = filename or fio.pathjoin(vars.workdir, 'config.yml')
-
-    if not utils.file_exists(filename) then
-        return nil, e_config_load:new('file %q does not exist', filename)
-    end
+local function validate_config(cwcfg, _)
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: Despite this api is local I want to be sure I've changed all callers properly. Guess I will remove it later.

File: cartridge/confapplier.lua [hunk 5]
```diff
@@ -16,362 +20,452 @@ local pool = require('cartridge.pool')
...
---- Load configuration from the filesystem.
--- Configuration is a YAML file.
--- @function load_from_file
+--- Validate configuration by all roles.
+-- @function validate_config
 -- @local
--- @tparam ?string filename Filename to load.
--- When omitted, the active configuration is loaded from `<workdir>/config.yml`.
--- @treturn[1] table
+-- @tparam table cwcfg_new
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: This is not to save the time, but to emphasize difference between good old `conf` (simple lua table) and `ClusterwideConfig` high-level abstraction

File: cartridge/confapplier.lua [hunk 6]
```diff
@@ -16,362 +20,452 @@ local pool = require('cartridge.pool')
...
-    if section_name == nil then
-        ret = vars.conf
-    else
-        ret = vars.conf[section_name]
+    local topology_cfg = cwcfg:get_readonly('topology') or {}
+    local box_opts = table.deepcopy(vars.box_opts)
+    box_opts.wal_dir = vars.workdir
+    box_opts.memtx_dir = vars.workdir
+    box_opts.vinyl_dir = vars.workdir
+    box_opts.listen = box.NULL
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: Yes, it does

File: cartridge/confapplier.lua [hunk 7]
```diff
@@ -16,362 +20,452 @@ local pool = require('cartridge.pool')
...
+        )
+        local leader_uuid = leaders_order[1]
+        local leader = topology_cfg.servers[leader_uuid]
+
+        if box_opts.instance_uuid == leader_uuid then
+            box_opts.replication = nil
+            box_opts.read_only = false
+        else
+            box_opts.replication = {pool.format_uri(leader.uri)}
+            -- box_opts.read_only = true
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: To allow overriding from env vars. (I'll remove it completely)

File: cartridge/confapplier.lua [hunk 8]
```diff
@@ -16,362 +20,452 @@ local pool = require('cartridge.pool')
...
+        BoxError:pcall(
+            box.schema.user.grant,
+            username, 'replication',
+            nil, nil, {if_not_exists = true}
+        )
     end
 
-    return roles.validate_config(conf_new, conf_old)
-end
+    do
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: The term 'replica' is meaningless at this moment of instance lifecycle. When instance goes RecoveringSnapshot there are no masters or replicas. They all look more like replicas.

Worry not, changing password doesn't cause replication conflict.

File: cartridge/confapplier.lua [hunk 9]
```diff
@@ -16,362 +20,452 @@ local pool = require('cartridge.pool')
...
-    local my_replicaset = conf.topology.replicasets[box.info.cluster.uuid]
-    local active_masters = topology.get_active_masters()
-    local is_master = false
-    if active_masters[box.info.cluster.uuid] == box.info.uuid then
-        is_master = true
-    end
+        -- TODO validate vshard groups
 
-    local is_rw = is_master or my_replicaset.all_rw
+        vars.cwcfg = cwcfg:lock()
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: No one will. This one will never be unlocked. You can only receive unlocked copy by calling `:copy()`

File: cartridge/topology.lua [hunk 10]
```diff
@@ -468,20 +475,30 @@ local function validate(topology_new, topology_old)
...
+
+    if topology_cfg.servers == nil then
+        return nil
     end
 
-    local advertise_uri = membership.myself().uri
-    for _it, instance_uuid, server in fun.filter(not_expelled, topology.servers) do
-        if server.uri == advertise_uri then
-            return instance_uuid, server.replicaset_uuid
+    for _it, instance_uuid, server in fun.filter(not_expelled, topology_cfg.servers) do
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/topology.lua]: In respect of https://luafun.github.io/basic.html

File: cartridge/topology.lua [hunk 11]
```diff
@@ -557,45 +569,23 @@ local function probe_missing_members(servers)
...
+local function get_fullmesh_replication(topology_cfg, replicaset_uuid)
+    checks('table', 'string')
+    assert(topology_cfg.__type ~= 'ClusterwideConfig')
+    assert(topology_cfg.servers ~= nil)
 
     local replication = {}
 
-    -- luacheck: ignore instance_uuid
-    for _it, instance_uuid, server in fun.filter(not_disabled, topology.servers) do
+    for _it, _, server in fun.filter(not_disabled, topology_cfg.servers) do
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/topology.lua]: I'm not a fan of luafun

File: cartridge/twophase.lua (hunk 1 above)
Comment (rosik) [reply cartridge/twophase.lua]: No, and it won't be. Abort is a consequence of preparation error, which sets `_2pc_error ~= nil`

P.S. I fixed preparation phase accordingly to ensure that nasty box.NULL doesn't break the algorithm

File: cartridge/twophase.lua (hunk 2 above)
Comment (rosik) [reply cartridge/twophase.lua]: Surely, yes. It keeps code sectioned.

File: cartridge/webui/api-topology.lua (hunk 3 above)
Comment (rosik) [reply cartridge/webui/api-topology.lua]: By mistake, reverted

File: cartridge/confapplier.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: But we are not afraid

File: cartridge/confapplier.lua [hunk 12]
```diff
@@ -16,362 +20,452 @@ local pool = require('cartridge.pool')
...
-    if not raw then
+    local ok, err = ddl_manager.validate_config(conf_new, conf_old)
+    if not ok then
         return nil, err
     end
 
-    local confdir = fio.dirname(filename)
+    return roles.validate_config(
+        cwcfg:get_readonly(),
+        vars.cwcfg and vars.cwcfg:get_readonly() or {}
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: Fixed

File: cartridge/admin.lua [hunk 13]
```diff
@@ -99,6 +99,13 @@ local function get_server_info(members, uuid, uri)
 end
 
 local function get_topology()
+    local state, err = confapplier.get_state()
+    if state == 'InitError'
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/admin.lua]: Fixed

File: cartridge.lua [hunk 14]
```diff
@@ -459,31 +455,13 @@ local function cfg(opts, box_opts)
 
     vshard_utils.set_known_groups(vshard_groups, opts.bucket_count)
 
-    bootstrap.set_box_opts(box_opts)
-    bootstrap.set_boot_opts({
+    local ok, err = confapplier.init({
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge.lua]: It does return boolean. Fixed the check

File: cartridge/clusterwide-config.lua [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+
+yaml.cfg({
+    encode_load_metatables = false,
+    decode_save_metatables = false,
+})
+
+local LoadConfigError = errors.new_class('LoadConfigError')
+local DecodeYamlError = errors.new_class('DecodeYamlError')
+
+local _cwcfg_mt
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/clusterwide-config.lua]: Well, besides underscore prefix (which is a default way to emphasize that variable is internal) and `_mt` suffix (which is common for Lua metatables), I want `cwcfg` (pronounced as `цэ вэ цэ фэ гэ`) to be a new cartridge term. It should be used to denote the complex high-level clusterwide config abstraction, which soon will replace simple Lua tables according to #329 RFC.

File: cartridge/failover.lua [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+end
+
+return {
+    cfg = cfg,
+
+    --- Get map of replicaset leaders.
+    -- @function get_active_leaders
+    -- @local
+    -- @return {[replicaset_uuid] = instance_uuid,...}
+    get_active_leaders = function()
```
Comment (knazarov) [inline cartridge/failover.lua]: don't write large functions inline

File: cartridge/roles.lua [hunk 17]
```diff
@@ -242,10 +243,9 @@ end
 -- @treturn[2] nil
 -- @treturn[2] table Error description
 local function validate_config(conf_new, conf_old)
-    if type(conf_new) ~= 'table'  then
-        return nil, ValidateConfigError:new('config must be a table')
-    end
     checks('table', 'table')
+    assert(conf_new.__type ~= 'ClusterwideConfig')
```
Comment (knazarov) [inline cartridge/roles.lua]: I'm puzzled a bit why you check that this isn't a `ClusterwideConfig`. At least write a comment what you are defending against.

File: cartridge/twophase.lua [hunk 18]
```diff
@@ -18,10 +19,11 @@ local pool = require('cartridge.pool')
 local utils = require('cartridge.utils')
 local topology = require('cartridge.topology')
 local confapplier = require('cartridge.confapplier')
+local ClusterwideConfig = require('cartridge.clusterwide-config')
```
Comment (knazarov) [inline cartridge/twophase.lua]: modules should be in snake_case

File: cartridge/twophase.lua (hunk 1 above)
Comment (rosik) [reply cartridge/twophase.lua]: No of course, and it will not be. Preparation stage remembers _2pc_error, and it will do `log.error('Clusterwide config update failed')`

File: cartridge/confapplier.lua [hunk 19]
```diff
@@ -16,362 +20,460 @@ local pool = require('cartridge.pool')
...
+    if vars.clusterwide_config then
+        conf_old = vars.clusterwide_config:get_readonly()
+    end
+    if conf_old == nil then
+        conf_old = {}
     end
 
-    local raw, err = utils.file_read(filename)
-    if not raw then
+    local ok, err = ddl_manager.validate_config(conf_new, conf_old)
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: No, both `ddl_manager` and `roles` validate_config are local and don't imply network calls.

Network calls are performed in `twophase` module. The chain is following: `twophase.patch_clusterwide -> netbox_call -> twophase.prepare_2pc -> confapplier.validate_config`

File: cartridge/confapplier.lua [hunk 20]
```diff
@@ -16,362 +20,460 @@ local pool = require('cartridge.pool')
...
+-- @function apply_config
+-- @local
+-- @tparam table conf
+-- @treturn[1] boolean true
+-- @treturn[2] nil
+-- @treturn[2] table Error description
+local function apply_config(clusterwide_config)
+    checks('ClusterwideConfig')
+    assert(clusterwide_config.locked)
+    assert(
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: > Why do you check it here?

To be sure I didn't make a mistake *before* calling `box.cfg`. It's just an assertion and I hope it will never raise.

File: cartridge/confapplier.lua [hunk 21]
```diff
@@ -16,362 +20,460 @@ local pool = require('cartridge.pool')
...
-        local ok, err = e_config_validate:pcall(
-            mod.validate_config, vars.conf, vars.conf
+    vars.instance_uuid = box.info.uuid
+    vars.replicaset_uuid = box.info.cluster.uuid
+    membership.set_payload('uuid', box.info.uuid)
+
+    if topology_cfg.servers == nil
+    or topology_cfg.servers[vars.instance_uuid] == nil
+    then
+        local err = BootError:new(
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/confapplier.lua]: > why not check it in advance?

I can't check it before box.cfg is recovered from snapshot. (of course I could parse snapshot by hands, but I don't want to do it anymore)

> how could it even happen at this stage?

Only manually, I guess (editing / copying clusterwide config by hands)
